src/clients/node/node.zig

```diff
@@ -266,36 +266,39 @@ fn on_completion(
     _ = timestamp;
 
     switch (packet.status) {
-        .ok => {},
+        .ok => {
+            switch (@as(Operation, @enumFromInt(packet.operation))) {
+                inline else => |op| {
+                    const event_count = @divExact(
+                        packet.data_size,
+                        @sizeOf(StateMachine.EventType(op)),
+                    );
+                    const buffer: BufferType(op) = .{
+                        .ptr = @ptrCast(packet),
+                        .count = event_count,
+                    };
+
+                    const Result = StateMachine.ResultType(op);
+                    const results: []const Result = @alignCast(std.mem.bytesAsSlice(
+                        Result,
+                        result_ptr.?[0..result_len],
+                    ));
+                    @memcpy(buffer.results()[0..results.len], results);
+
+                    packet.data_size = @bitCast(BufferSize{
+                        .event_count = @intCast(event_count),
+                        .result_count = @intCast(results.len),
+                    });
+                },
+                .pulse => unreachable,
+            }
+        },
```

Comment:
It isn't immediately obvious that `defer buffer.free();` remains correct in `on_completion_js`, but I think it does. We don't touch `event_count` part of `packet.data_size`, and that tells us what we should be freeing.